client/src/main/java/org/asynchttpclient/cookie/ThreadSafeCookieStore.java

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (c) 2017 AsyncHttpClient Project. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at
+ *     http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package org.asynchttpclient.cookie;
+
+import io.netty.handler.codec.http.cookie.Cookie;
+import org.asynchttpclient.uri.Uri;
+
+import java.util.*;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+public class ThreadSafeCookieStore implements CookieStore {
+
+    private Map<CookieKey, StoredCookie> cookieJar = new HashMap<>();
+    private ReentrantLock lock = new ReentrantLock();
+
+    @Override
+    public void add(Uri uri, Cookie cookie) {
+        String thisRequestDomain = requestDomain(uri);
+        String thisRequestPath = requestPath(uri);
+
+        try {
+            lock.lock();
+            add(thisRequestDomain, thisRequestPath, cookie);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public List<Cookie> get(Uri uri) {
+        try {
+            lock.lock();
+            return get(requestDomain(uri), requestPath(uri), uri.isSecured());
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public List<Cookie> getAll() {
+        try {
+            lock.lock();
+            final boolean[] removeExpired = {false};
+            List<Cookie> result = cookieJar
+                    .entrySet()
+                    .parallelStream()
+                    .filter(pair -> {
+                        boolean hasCookieExpired = hasCookieExpired(pair.getValue().cookie, pair.getValue().createdAt);
+                        if (hasCookieExpired && !removeExpired[0])
+                            removeExpired[0] = true;
+                        return !hasCookieExpired;
+                    })
+                    .map(pair -> pair.getValue().cookie)
+                    .collect(Collectors.toList());
+
+            if (removeExpired[0])
+                removeExpired();
+
+            return result;
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public boolean remove(Cookie cookie) {
+        try {
+            lock.lock();
+            return cookieJar.entrySet().removeIf(v -> v.getValue().cookie == cookie);
```

Comment:
No, I intentionally wanted to remove by reference because intention is to modify cookie jar using `Cookie` objects returned by `.getAll()` or `get()` methods. This is similar to what `java.net.InMemoryCookieStore.remove()` does.

Reply:
I think this is specific to your use case, Other people might want a different strategy, eg drop all cookies for a given domain. So it might be best to change the parameter to a `Predicate<Cookie>`.